specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,248 @@
+            words
+        default_refinement_kwargs: dict, optional
+            words
+        """
+        self.input_spectrum = input_spectrum
+        self.model = model
+        self.spectral_unit = spectral_unit
+        self.fitter = fitter
+        self.default_refinement_method = default_refinement_method
+        self.default_refinement_kwargs = default_refinement_kwargs
+
+        if self.default_refinement_method in ("gaussian", "min", "max"):
+            if 'range' not in self.default_refinement_kwargs:
+                raise ValueError("You must define 'range' in default_refinement_kwargs to use "
+                                 f"{self.refinement_method} refinement.")
+        elif (self.default_refinement_method == "gradient" and 'direction' not in
+                self.default_refinement_kwargs):
+            raise ValueError("You must define 'direction' in default_refinement_kwargs to use "
+                             "gradient refinement")
+
+        self.lines = []
+        if isinstance(lines, str):
+            if lines in self.available_line_lists:
+                raise ValueError(f"Line list '{lines}' is not an available line list.")
+        else:
+            for line in lines:
+                if isinstance(line, CalibrationLine):
+                    self.lines.append(line)
+                else:
+                    self.lines.append(CalibrationLine(self.input_spectrum, line[0], line[1],
+                                      self.default_refinement_method,
+                                      self.default_refinement_kwargs))
+
+    @classmethod
+    def autoidentify(cls, input_spectrum, line_list, model=Linear1D):
+        # line_list could be a string ("common stellar") or an object
+        # this builds a list of CalibrationLine objects and passes to __init__
+        return cls(...)
+
+    @property
+    def refined_lines(self):
+        return [line.with_refined_pixel for line in self.lines]
+
+    @property
+    def refined_pixels(self):
+        # useful for plotting over spectrum to change input lines/refinement options
+        return [(line.wavelength, line.refined_pixel) for line in self.lines]
+
+    @cached_property
+    def wcs(self):
+        # computes and returns WCS after fitting self.model to self.refined_pixels
+        x = [line[1] for line in self.refined_pixels] * u.pix
+        y = [line[0].value for line in self.refined_pixels] * self.spectral_unit
+
+        # Fit the model
+        self.model = self.fitter(self.model, x, y)
+
+        # Build a GWCS pipeline from the fitted model
+        pixel_frame = cf.CoordinateFrame(1, "SPECTRAL", [0,], axes_names=["x",], unit=[u.pix,])
+        spectral_frame = cf.SpectralFrame(axes_names=["wavelength",], unit=[self.spectral_unit,])
+
+        pipeline = [(pixel_frame, self.model), (spectral_frame, None)]
+
+        wcsobj = wcs.WCS(pipeline)
+
+        return wcsobj
+
+    def apply_to_spectrum(self, spectrum=None):
```

Comment:
Do we want `__call__` instead, to just call this, to do something else, or to not exist?  I think we will eventually want calling to do _something_ since that pattern is supported through most/all of the other specreduce operations.

Reply:
I remembered @eteq was somewhat against having a `__call__` at all, so I had left it out for now. If it's more consistent with the way the rest of the package works, then I guess we can think about what makes the most sense. Having it call this method on the default (input) spectrum seems like an ok behavior to me, but I don't feel strongly about it.

Reply:
Ok, I think we'll want `__call__` to do _something_ eventually, but I'm fine deferring that until the rest of the API begins to converge in follow-up efforts.

Reply:
a workflow we will need to support is where we iterate the fitting process. one would start with a small number of matched lines and a simple model, use the model to match more lines, re-fit, tweak model, match more lines, re-fit/tweak/match, and repeat until a convergence criteria is met. my view is that this class/function should be within that loop, but not contain that loop. i don't see where a ``__call__`` fits into that scheme and i think a `apply_to_spectrum()` method is a clearer way of applying the results to a spectrum.